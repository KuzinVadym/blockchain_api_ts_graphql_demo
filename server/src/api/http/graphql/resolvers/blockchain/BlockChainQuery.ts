import { Arg, Ctx, Int, Query, Resolver } from 'type-graphql';
import { withHttpService } from '../../../shared/utils/withHttpService';
import { BlockType } from './BlockType';
import { RawBlockType } from './RawBlockType';

@Resolver((of) => BlockType)
export class BlockChainQuery {
  @Query(() => [BlockType])
  public async blocks(@Ctx() ctx): Promise<BlockType[]> {
    const blockChainHttpService = withHttpService('blockchain', ctx.getState);
    return blockChainHttpService.getBlocks();
  }

  @Query(() => RawBlockType)
  public async rawBlock(
    @Arg('hash', (type) => String) hash: string,
    @Ctx() ctx
  ): Promise<RawBlockType> {
    const blockChainHttpService = withHttpService('blockchain', ctx.getState);
    return blockChainHttpService.getRawBlock(hash);
  }
}
