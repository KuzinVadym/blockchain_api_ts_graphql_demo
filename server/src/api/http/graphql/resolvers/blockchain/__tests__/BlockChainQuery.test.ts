import 'reflect-metadata';
import {
  ApolloServerTestClient,
  createTestClient
} from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import { buildSchema } from 'type-graphql';
import { BlockChainQuery } from '../BlockChainQuery';

const contextMock = {
  getState: jest.fn()
};

const blockChainServiceMock = {
  getBlocks: jest.fn(),
  getRawBlock: jest.fn()
};

describe('CommitQuery', () => {
  let server: ApolloServer;

  beforeAll(async () => {
    const schema = await buildSchema({ resolvers: [BlockChainQuery] });
    server = new ApolloServer({
      schema,
      context: contextMock
    });
  });

  beforeEach(async () => {
    blockChainServiceMock.getBlocks.mockResolvedValue([
      {
        hash:
          '000000000000000000082bac0859c43f89c35423049712658ee9a0963c15aaf3',
        time: 1614211561,
        height: 672046
      }
    ]);
    blockChainServiceMock.getRawBlock.mockResolvedValue({
      size: 1819006,
      block_index: 0
    });

    contextMock.getState.mockReturnValue({
      httpServices: {
        blockchain: blockChainServiceMock
      }
    });
  });

  afterEach(() => {
    blockChainServiceMock.getBlocks.mockReset();
    blockChainServiceMock.getRawBlock.mockReset();
  });

  it('returns blocks data', async () => {
    const testQuery = `{
      blocks {
        hash
        time
        height
      }
    }
  `;

    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: {} });
    expect(res).toMatchSnapshot();
  });

  it('returns Raw Block data', async () => {
    const testQuery = `
    query BlockChainQuery($hash: String!){
      rawBlock(hash: $hash) {
        size
        block_index
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: { hash: '1' } });
    expect(res).toMatchSnapshot();
  });

  it('cals service with right params', async () => {
    const testQuery = `
    query BlockChainQuery($hash: String!){
      rawBlock(hash: $hash) {
        size
        block_index
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    await query({ query: testQuery, variables: { hash: '1' } });

    expect(blockChainServiceMock.getRawBlock).toBeCalledWith('1');
  });
});
