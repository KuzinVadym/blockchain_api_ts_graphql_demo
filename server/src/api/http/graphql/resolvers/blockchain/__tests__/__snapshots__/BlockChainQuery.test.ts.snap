// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitQuery returns Raw Block data 1`] = `
Object {
  "data": Object {
    "rawBlock": Object {
      "block_index": 0,
      "size": 1819006,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`CommitQuery returns blocks data 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "hash": "000000000000000000082bac0859c43f89c35423049712658ee9a0963c15aaf3",
        "height": 672046,
        "time": 1614211561,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
